### PR TITLE
fix: better locking in auth concurrent operations

### PR DESCRIPTION
**Pull request by bastiandoetsch**
### Description

_Provide description of this PR and changes, if linked Jira ticket doesn't cover it in full._

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [ ] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced

### CODE REVIEW

File: application/server/server.go [hunk 1]
```diff
@@ -638,13 +638,19 @@ func textDocumentDidOpenHandler() jrpc2.Handler {
 		logger := c.Logger().With().Str("method", "TextDocumentDidOpenHandler").Str("documentURI", filePath).Logger()
 
 		logger.Info().Msg("Receiving")
-		folder := workspace.Get().GetFolderContaining(filePath)
+		folder := c.Workspace().GetFolderContaining(filePath)
```
Comment (ShawkyZ) [inline application/server/server.go:641]: let's pass the config to this eventHandler 
maybe we can get rid of all config.CurrentConfig all together here

File: application/server/server.go [hunk 2]
```diff
@@ -490,23 +490,23 @@ func startOfflineDetection(c *config.Config) { //nolint:unused // this is gonna
...
 	for {
-		deleteExpiredCache()
+		deleteExpiredCache(c)
 		time.Sleep(time.Duration(persistence.ExpirationInSeconds) * time.Second)
 	}
 }
 
-func addWorkspaceFolders(c *config.Config, params types.InitializeParams, w *workspace.Workspace) {
+func addWorkspaceFolders(c *config.Config, params types.InitializeParams, w types.Workspace) {
 	const method = "addWorkspaceFolders"
```
Comment (ShawkyZ) [inline application/server/server.go:510]: `w types.Workspace` argument isn't needed anymore here we can get it from config

File: domain/ide/command/clear_cache.go [hunk 3]
```diff
@@ -77,7 +77,7 @@ func (cmd *clearCache) Execute(_ context.Context) (any, error) {
 }
 
 func (cmd *clearCache) purgeInMemoryCache(logger *zerolog.Logger, folderUri *lsp.DocumentURI) {
-	ws := workspace.Get()
+	ws := config.CurrentConfig().Workspace()
```
Comment (ShawkyZ) [inline domain/ide/command/clear_cache.go:80]: we can use `cmd.c` here since u injected it

File: domain/ide/command/clear_cache.go [hunk 4]
```diff
@@ -93,9 +93,9 @@ func (cmd *clearCache) purgeInMemoryCache(logger *zerolog.Logger, folderUri *lsp
 
 func (cmd *clearCache) purgePersistedCache(logger *zerolog.Logger, folderUri *lsp.DocumentURI) {
 	var folderList []string
-	ws := workspace.Get()
-	scanPersister := ws.ScanPersister()
-	if scanPersister == nil {
+	ws := cmd.c.Workspace()
+	clearerExister := ws.GetScanSnapshotClearerExister()
```
Comment (ShawkyZ) [inline domain/ide/command/clear_cache.go:97]: wdyt? 
```suggestion
	clearerExister := ws.GetScanSnapshotPersister()
```

File: domain/ide/workspace/workspace.go [hunk 5]
```diff
@@ -33,13 +33,10 @@ import (
...
 )
 
-// todo can we do without a singleton?
-var instance *Workspace
-
 // Workspace represents the highest entity in an IDE that contains code. A workspace may contain multiple folders
 type Workspace struct {
 	mutex               sync.Mutex
-	folders             map[string]*Folder
+	folders             map[string]types.Folder
```
Comment (ShawkyZ) [inline domain/ide/workspace/workspace.go:39]: shouldn't this be a pointer?

Comment (ShawkyZ) [reply domain/ide/workspace/workspace.go:39]: it's an interface.. nvm

File: infrastructure/authentication/auth_service_impl.go [hunk 6]
```diff
@@ -67,10 +67,20 @@ func NewAuthenticationService(c *config.Config, authProviders AuthenticationProv
 }
 
 func (a *AuthenticationServiceImpl) Provider() AuthenticationProvider {
+	a.m.Lock()
```
Comment (ShawkyZ) [inline infrastructure/authentication/auth_service_impl.go:70]:
```suggestion
	a.m.RLock()
```

File: infrastructure/authentication/auth_service_impl.go [hunk 7]
```diff
@@ -67,10 +67,20 @@ func NewAuthenticationService(c *config.Config, authProviders AuthenticationProv
...
+	defer a.m.Unlock()
+
 	return a.provider
 }
 
 func (a *AuthenticationServiceImpl) Authenticate(ctx context.Context) (token string, err error) {
+	a.m.Lock()
```
Comment (ShawkyZ) [inline infrastructure/authentication/auth_service_impl.go:77]: same here

File: infrastructure/authentication/auth_service_impl.go [hunk 8]
```diff
@@ -179,29 +190,37 @@ func (a *AuthenticationServiceImpl) UpdateCredentials(newToken string, sendNotif
...
-	a.UpdateCredentials("", true)
-	a.ConfigureProviders(a.c)
+	a.updateCredentials("", true)
+	a.configureProviders(a.c)
 }
 
 // IsAuthenticated returns true if the token is verified
 // If the token is set, but not valid IsAuthenticated returns false
 func (a *AuthenticationServiceImpl) IsAuthenticated() bool {
+	a.m.Lock()
```
Comment (ShawkyZ) [inline infrastructure/authentication/auth_service_impl.go:212]: Readlock ?

**Review by ShawkyZ — APPROVED** [review]

File: application/server/server.go (hunk 2 above)
Comment (bastiandoetsch) [reply application/server/server.go:510]: true

File: application/server/server.go (hunk 1 above)
Comment (bastiandoetsch) [reply application/server/server.go:641]: ok

**Pull request merged**